full_build_upgraded/full_build/ops_security/observability.py

```
@@ -35,4 +110,63 @@ def record_error(path: str, error: str) -> None:
+
+def record_selfloop_execution(
+    iterations: int,
+    confidence: float,
+    duration: float,
+    model_tier: str
+) -> None:
+    """Record metrics for a self-loop execution.
+
+    Args:
+        iterations: Number of self-loop iterations.
+        confidence: Confidence score (0.0–1.0).
```

Ruff (0.13.3)

Docstring contains ambiguous `–` (EN DASH). Did you mean `-` (HYPHEN-MINUS)?

(RUF002)